QueryParser.cpp

```diff
@@ -292,16 +302,22 @@ Query transform(const parse_tree::node &n) {
 
 Command transform_command(const parse_tree::node &n) {
     if (n.is<select>()) {
-        std::set<std::string> taints;
-        parse_tree::node *expr = &*n.children[0];
 
-        if (expr->is<with_taints>()) {
-            for (const auto &taint : n.children[1]->children) {
-                taints.insert(transform_string(*taint));
+        int iter = 0;
+        std::set<std::string> taints;
+        while (true) {
+            if (n.children[iter]->is<with_taints_token>()) {
+                for (const auto &taint : n.children[iter + 1]->children) {
+                    taints.insert(transform_string(*taint));
+                }
+                iter += 2;
+            }
+            if (n.children[iter]->is<select_body>()) {
+                const auto &body = n.children[iter];
+                const auto &expr = body->children[0];
+                return Command(SelectCommand(transform(*expr), taints));
```

Comment:
This loops seems a bit confusing, at least for me.
Wouldn't it be more readable if it were structured like:
```
for ( ;is<with_taints_token>; iter += 2) {
  add tain
}
// As the parser enforces tree structure, iter should point to select_body
return Command(...)
```

Reply:
Nah, that's not the goal here. It may not be clear, because it's done in anticipation of change that's not there yet (also, there's only one taint expression possible, so there's no point on looping with it).

But we plan to add more modifiers, so that for example:

```
select with taints ["kot", "hmm"] "xyz";
select into iterator "kot";  # syntax WIP 
select with taints ["kot", "hmm"] into iterator "kot";
```

can all both. In other words, iter will point at the next token (select_body, with_taints_token, into_iterator_token) at all times, so that the for loop can dispatch it correctly.

But if you have idea how to improve it, feel free to suggest. You're right that select_body is always last so I can factor that out of the loop - will it help?

Reply:
I've chnaged this a bit in the new version, PTAL now

Reply:
Looks better, and I appreciate the comments :)